Skip excess keys on serialization load

johnswanson opened the pull request:
When loading from a backup, we should not fail if the previous version saved keys that are no longer needed in the current version. This adds a step (after ingestion but before loading) that removes any "extra" keys from the ingested entity.

Co-authored by: escherize <343288+escherize@users.noreply.github.com>

johnswanson committed:
Skip excess keys on serialization load

When loading from a backup, we should not fail if the previous version saved keys that are no longer needed in the
current version. This adds a step (after ingestion but before loading) that removes any "extra" keys from the ingested
entity.

Co-authored by: bryan <343288+escherize@users.noreply.github.com>

johnswanson committed:
Try using `memoize-for-application-db`

I'm not entirely sure, but I think *possibly* the issue was that tests used a different application-db than the
development environment, which caused memoization to fail.

I'd like to look into this more to confirm. Frustratingly, I was able to run `(do (dev) (start!))` and get a failure
earlier, but now I can't reproduce :facepalm:. I wonder if possibly there is some interaction I didn't think about -
e.g. did I run a test before running the above, causing the function to get called with a different app-db, causing
the (wrong) result to get stored?

johnswanson committed:
Fix drop-excess-keys

There are two changes here:

- first, call `load-xform-basics` at the *end* of `load-xform` for Collections instead of the beginning. Collections
have a field, `parent_id`, that is translated internally into a `location`. When `load-xform-basics` removes keys that
aren't columns in the database, we remove `parent_id` and the Collection ends up with a null location. Note that the
docstring for `load-xform-basics` previously suggested that it should be called at the *beginning* of any custom
implementation of `load-xform` - but there's really no need for that, as long as we do call it. I updated the
docstring accordingly.

- second, add a new multimethod, `serdes/ingested-model-columns`. The default implementation uses the same method
we've been using - looks in the application DB and returns the set of keys that correspond to the columns there.
However, the multimethod can be overridden in cases like "Action", where the model corresponds to more than one table
in the database (e.g. an Action is actually a row in both the `action` table and one of the `query_action`,
`http_action`, or `implicit_action` tables).

johnswanson committed:
fix tests for loading with extra keys

There were two problems with the existing code:

- running the test left the database in a state that caused subsequent tests to fail :sob:

- running the test _without_ first running `dev/start!` would fail

This led to ... extreme confusion, as I'd try to rerun other failing tests and they'd pass, then try running these
tests and they'd fail... or sometimes pass, depending on if I'd run `start!` from my REPL or not... and then try
running the original failing tests again and they'd fail. Just a lot of degrees of freedom, leading to confusion.

The new test starts with a completely blank database, sets it up, and ensures that we can run
`ensure-audit-db-installed!` without error.

darksciencebase commented:
cc @piranha

piranha commented on src/metabase/models/serialization.clj:
```diff
@@ -507,15 +509,52 @@
   This is a mirror (but not precise inverse) of [[extract-one-basics]]."
   [ingested]
-  (dissoc ingested :serdes/meta))
+  (-> ingested drop-excess-keys (dissoc :serdes/meta)))
```
Feels like `:serdes/meta` will be removed by `drop-excess-keys` anyways?

johnswanson replied:
Good point, thank you!

piranha commented on src/metabase/util/connection.clj:
```diff
@@ -0,0 +1,21 @@
+              (repeatedly
+               (fn []
+                 (when (.next rset)
+                   [(.getString rset "COLUMN_NAME") (.getString rset "TYPE_NAME")]))))))))
```
IDK if you're into that (I'm ok with code in place here), but since 1.11 there is `clojure.core/iteration`:

```
(into {} (iteration (fn [_] (when (.next cs)
                                  [(.getString cs "COLUMN_NAME") (.getString cs "TYPE_NAME")]))))
```

piranha replied:
Ah, you've just moved the code :)

johnswanson replied:
Ah yeah, I did just move it, but also TIL about `iteration` - very nice! 🎉

piranha commented:
The code is fine, but I'm not sold on having a big e2e test just to check this functionality. Maybe some smaller unit-test in somewhere in `v2.load-test` will suffice, what do you think?

johnswanson committed:
CR response

- smaller unit test for functionality

- `clojure.core/iteration` :tada:

- remove unnecessary `(dissoc :serdes/meta)`

sansolovyov reviewed (approved):
Cool!
